frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -1046,67 +1066,86 @@ func dispatchCopyFileOp(d *dispatchState, c instructions.SourcesAndDest, sourceS
 		return err
 	}
 
+	name := uppercaseCmd(processCmdEnv(cfg.opt.shlex, cfg.cmdToPrint.String(), env))
 	fileOpt := []llb.ConstraintsOpt{
-		llb.WithCustomName(prefixCommand(d, uppercaseCmd(processCmdEnv(opt.shlex, cmdToPrint.String(), env)), d.prefixPlatform, &platform, env)),
-		location(opt.sourceMap, loc),
+		llb.WithCustomName(prefixCommand(d, name, d.prefixPlatform, &platform, env)),
+		location(cfg.opt.sourceMap, cfg.location),
 	}
 	if d.ignoreCache {
 		fileOpt = append(fileOpt, llb.IgnoreCache)
 	}
 
-	d.state = d.state.File(a, fileOpt...)
+	if cfg.opt.llbCaps.Supports(pb.CapMergeOp) == nil && cfg.link && cfg.chmod == "" {
+		mergeOpt := append(fileOpt, llb.WithCustomName(prefixCommand(d, "LINK "+name, d.prefixPlatform, &platform, env)))
+		d.state = llb.Merge([]llb.State{d.state, llb.Scratch().File(a, fileOpt...)}, mergeOpt...)
```

Comment:
added this

Reply:
Thanks, though I just thought of something that's worth considering. I remembered that this behavior implies that any deletions present in the layers of the source of the link will be applied on top of the new base, which is different than the behavior you get with copies where all the source layers are squashed and deletions are discarded.

I can't think of a single realistic Dockerfile use case where this would matter, but did want to bring it up in case you think this somewhat divergent behavior might matter.

Reply:
I'm not sure about the whiteouts but thought about another case. This currently would mess up the history array in the image as history array will only expect that one layer was generated by the command. Possibly something can be done with adding some markers to overcome this (not quite sure) but I think I will revert this part for now.

Reply:
Okay cool sgtm, the use case for this optimization was pretty niche anyways so doesn't hurt much to save it for the future.